From dbf1c833f87b2132e978b7ea46cc02a189a136d4 Mon Sep 17 00:00:00 2001
From: Kyle Hsiao <kylehsiao@google.com>
Date: Mon, 11 May 2026 12:11:33 +0000
Subject: [PATCH 08/13] Prevent information disclosure over RF in rw_t5t.cc

Port missing bounds checks from b/143109193, b/143155861, and b/143106535 to rw_t5t.cc. Verify response buffer length before copying stale GKI buffer slack into outgoing WRITE_SINGLE_BLOCK commands.

Bug: 508389908
Test: R/W T5T tags
Flag: EXEMPT BUGFIX
(cherry picked from commit 5fdacfff0f153d63369252167961c1c6effb4a61)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:104a741a53ac7b4ab685236ca2da2ca969c85d3d
Merged-In: Ib8c54a103d20df76652df40797b39144b38b3d70
Change-Id: Ib8c54a103d20df76652df40797b39144b38b3d70
---
 libnfc-nci/src/nfc/tags/rw_t5t.cc | 19 ++++++++++++++-----
 1 file changed, 14 insertions(+), 5 deletions(-)

diff --git a/libnfc-nci/src/nfc/tags/rw_t5t.cc b/libnfc-nci/src/nfc/tags/rw_t5t.cc
index 1ea262b..1f4b8ec 100644
--- a/libnfc-nci/src/nfc/tags/rw_t5t.cc
+++ b/libnfc-nci/src/nfc/tags/rw_t5t.cc
@@ -571,8 +571,11 @@ void rw_t5t_sm_update_ndef(NFC_HDR* p_resp) {
 
       block_number = (p_i93->ndef_tlv_start_offset + 1) / p_i93->block_size;
 
-      if (rw_i93_send_cmd_write_single_block(block_number, p) ==
-          NFC_STATUS_OK) {
+      if (length < p_i93->block_size) {
+        android_errorWriteLog(0x534e4554, "508389908");
+        rw_i93_handle_error(NFC_STATUS_FAILED);
+      } else if (rw_i93_send_cmd_write_single_block(block_number, p) ==
+                 NFC_STATUS_OK) {
         /* update next writing offset */
         p_i93->rw_offset = (block_number + 1) * p_i93->block_size;
         p_i93->sub_state = RW_I93_SUBSTATE_WRITE_NDEF;
@@ -749,8 +752,11 @@ void rw_t5t_sm_update_ndef(NFC_HDR* p_resp) {
 
           block_number = (p_i93->rw_offset / p_i93->block_size);
 
-          if (rw_i93_send_cmd_write_single_block(block_number, p) ==
-              NFC_STATUS_OK) {
+          if (length < p_i93->block_size) {
+            android_errorWriteLog(0x534e4554, "508389908");
+            rw_i93_handle_error(NFC_STATUS_FAILED);
+          } else if (rw_i93_send_cmd_write_single_block(block_number, p) ==
+                     NFC_STATUS_OK) {
             /* set offset to the beginning of next block */
             p_i93->rw_offset +=
                 p_i93->block_size - (p_i93->rw_offset % p_i93->block_size);
@@ -831,7 +837,10 @@ void rw_t5t_sm_set_read_only(NFC_HDR* p_resp) {
       /* mark CC as read-only */
       *(p + 1) |= I93_ICODE_CC_READ_ONLY;
 
-      if (rw_i93_send_cmd_write_single_block(0, p) == NFC_STATUS_OK) {
+      if (length < p_i93->block_size) {
+        android_errorWriteLog(0x534e4554, "508389908");
+        rw_i93_handle_error(NFC_STATUS_FAILED);
+      } else if (rw_i93_send_cmd_write_single_block(0, p) == NFC_STATUS_OK) {
         p_i93->sub_state = RW_I93_SUBSTATE_WAIT_UPDATE_CC;
       } else {
         rw_i93_handle_error(NFC_STATUS_FAILED);
-- 
2.55.0

